[GSoC] Compatibility Changes in Trial Controller

Electronic-Waste opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, check our contributor guidelines https://www.kubeflow.org/docs/about/contributing
2. To know more about Katib components, check developer guide https://github.com/kubeflow/katib/blob/master/docs/developer-guide.md
3. If you want *faster* PR reviews, check how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:

I made some compatibility changes to the Trial Controller. Design details: https://github.com/kubeflow/katib/blob/master/docs/proposals/push-based-metrics-collection.md#compatibility-changes-in-trial-controller

**Which issue(s) this PR fixes** _(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)_:
Fixes #

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/katib/) included if any changes are user facing

Electronic-Waste commented:
@andreyvelich @johnugeorge PTAL👀 if you are available. Thanks!

ref issue: #2340

andreyvelich commented on pkg/controller.v1beta1/trial/trial_controller.go:
```diff
@@ -244,11 +245,15 @@ func (r *ReconcileTrial) reconcileTrial(instance *trialsv1beta1.Trial) error {
 		if jobStatus.Condition == trialutil.JobSucceeded && instance.Status.Observation == nil {
-			logger.Info("Trial job is succeeded but metrics are not reported, reconcile requeued")
-			return errMetricsNotReported
+			if instance.Spec.MetricsCollector.Collector == nil ||
```
Do we have a use-case when collector is nil ? I think, we add the StdOut as a default MC: https://github.com/kubeflow/katib/blob/4b477e82d56433389a70da2ffc54088ebc9c449d/pkg/apis/controller/experiments/v1beta1/experiment_defaults.go#L131

Electronic-Waste replied:
@andreyvelich It seems that we pass a nil collector in the testcase: https://github.com/kubeflow/katib/blob/51b246fa1c23db130c373fa9749897a0d0b334ab/pkg/controller.v1beta1/trial/trial_controller_test.go#L102-L124

And I got seg fault in CI/CD when I did not add nil condition judgement: 

<img width="1083" alt="截屏2024-08-02 14 59 35" src="https://github.com/user-attachments/assets/4ddb2d23-2d0f-4626-b77e-c1178553250b">

Electronic-Waste replied:
I'll investigate into this question and fix CI/CD.

Electronic-Waste replied:
@andreyvelich I find the nil condition occurred due to: https://github.com/kubeflow/katib/blob/51b246fa1c23db130c373fa9749897a0d0b334ab/pkg/controller.v1beta1/trial/trial_controller_test.go#L425-L445

which omitted the `MetricsCollector` field when creating Trials. I have added it now.

andreyvelich replied:
Amazing, thank you!

andreyvelich commented on pkg/controller.v1beta1/trial/trial_controller.go:
```diff
@@ -244,11 +245,15 @@ func (r *ReconcileTrial) reconcileTrial(instance *trialsv1beta1.Trial) error {
-			logger.Info("Trial job is succeeded but metrics are not reported, reconcile requeued")
-			return errMetricsNotReported
+			if instance.Spec.MetricsCollector.Collector == nil ||
+				instance.Spec.MetricsCollector.Collector.Kind != commonapiv1beta1.PushCollector {
```
Can you just add this condition to the L251 ? E.g.
```golang
if jobStatus.Condition == trialutil.JobSucceeded && instance.Status.Observation == nil && instance.Spec.MetricsCollector.Collector.Kind != commonapiv1beta1.PushCollector
```

Electronic-Waste replied:
Okay for that.

andreyvelich commented on pkg/controller.v1beta1/trial/trial_controller_util.go:
```diff
@@ -70,6 +70,14 @@ func (r *ReconcileTrial) UpdateTrialStatusCondition(instance *trialsv1beta1.Tria
 			reason := TrialMetricsUnavailableReason
 
+			// If the type of metrics collector is Push, We should insert an unavailable value to Katib DB
+			if instance.Spec.MetricsCollector.Collector != nil &&
```
This might not be needed.

andreyvelich commented on pkg/controller.v1beta1/trial/trial_controller_util.go:
```diff
@@ -70,6 +70,14 @@ func (r *ReconcileTrial) UpdateTrialStatusCondition(instance *trialsv1beta1.Tria
 			reason := TrialMetricsUnavailableReason
 
+			// If the type of metrics collector is Push, We should insert an unavailable value to Katib DB
+			if instance.Spec.MetricsCollector.Collector != nil &&
```
We might need to add an integration test for it: https://github.com/kubeflow/katib/blob/4b477e82d56433389a70da2ffc54088ebc9c449d/pkg/controller.v1beta1/trial/trial_controller_test.go
cc @tenzen-y

andreyvelich replied:
Additionally, please try to run the example with Push collector when you don't insert objective metrics and verify that Trial has been moved into `MetricsUnavailable` status.

Electronic-Waste replied:
> We might need to add an integration test for it: https://github.com/kubeflow/katib/blob/4b477e82d56433389a70da2ffc54088ebc9c449d/pkg/controller.v1beta1/trial/trial_controller_test.go
cc @tenzen-y

@andreyvelich If I understand correctly, do you mean we need to add an integration test for Push Collector?

> Additionally, please try to run the example with Push collector when you don't insert objective metrics and verify that Trial has been moved into MetricsUnavailable status.

SGTM.

Electronic-Waste replied:
@andreyvelich I tried to run the example with `report_metrics` commented. And the result shows that all the Trials have been moved into `MetricsUnavailable` status.

<img width="966" alt="截屏2024-08-02 16 36 35" src="https://github.com/user-attachments/assets/fc1ec4e3-db5e-476c-9d6c-a6d6b77a5902">

andreyvelich replied:
>@andreyvelich If I understand correctly, do you mean we need to add an integration test for Push Collector?

Yes, just to verify that appropriate functional calls are happening.

Electronic-Waste replied:
Okay, I'll implement that in the next few commits.

Electronic-Waste replied:
The integration test for Push Collector has been added now. Since there are no differences for MCs in successful scenario, I only add a test case in which the batch job is completed but no metrics are reported to the Katib DB.

Does it meet our need? WDYT👀 @andreyvelich

andreyvelich replied:
Yes, I think, it should be good. Thank you!

andreyvelich reviewed (commented):
Thank you for this @Electronic-Waste!
cc @kubeflow/wg-automl-leads

andreyvelich commented:
/rerun-all

andreyvelich commented on pkg/controller.v1beta1/trial/trial_controller_util.go:
```diff
@@ -162,6 +169,23 @@ func (r *ReconcileTrial) updateFinalizers(instance *trialsv1beta1.Trial, finaliz
 	}
 }
 
+func (r *ReconcileTrial) reportUnavailableMetrics(instance *trialsv1beta1.Trial) error {
```
Please can you verify if user doesn't report any results for `additionalMetricNames` it inserts the `unavailable` metrics to the observation.
For the pull-based metrics, we are doing this as part of this script: https://github.com/kubeflow/katib/blob/master/pkg/controller.v1beta1/trial/trial_controller_util.go#L170-L176.
So the observation (objective metric is `result`) might look as follows:
```
  observation:
    metrics:
    - latest: "123"
      max: "123"
      min: "123"
      name: result
    - latest: unavailable
      max: unavailable
      min: unavailable
      name: metrics1
    - latest: unavailable
      max: unavailable
      min: unavailable
      name: metrics2
```

Electronic-Waste replied:
Okay, I'll investigate into this.

Electronic-Waste replied:
@andreyvelich  The result shows that Katib will insert the `unavailable` metrics into the observation when the user doesn't report any results for `additionalMetricsNames`. This logic is not implemented by Push Collector.

<img width="1046" alt="截屏2024-08-15 17 40 51" src="https://github.com/user-attachments/assets/d9ae3788-80d0-4949-ad2e-e7704e873063">

<img width="575" alt="截屏2024-08-15 17 41 15" src="https://github.com/user-attachments/assets/50e1579c-550c-4c62-a648-d52892004960">

tenzen-y commented:
> /assign @johnugeorge @tenzen-y Please review it when you have time.

ACK

tenzen-y commented on pkg/controller.v1beta1/trial/trial_controller_test.go:
```diff
@@ -290,14 +290,47 @@ func TestReconcileBatchJob(t *testing.T) {
+		g := gomega.NewGomegaWithT(t)
+		gomock.InOrder(
+			mockManagerClient.EXPECT().GetTrialObservationLog(gomock.Any()).Return(observationLogUnavailable, nil).MinTimes(1),
+			mockManagerClient.EXPECT().ReportTrialObservationLog(gomock.Any(), gomock.Any()).Return(nil, nil).MinTimes(1),
```
What happens if this reporting fails?

Electronic-Waste replied:
It will fail to insert `Unavailable` to Katib DB. I haven't considered this scenario. Do you mean I need to add another UT for the failed case?

andreyvelich replied:
If we can the unit test for it, that would be great @Electronic-Waste!

Electronic-Waste replied:
@andreyvelich I'll add the UT when I have the bandwidth this two weeks.

tenzen-y reviewed (commented):
Basically, lgtm

Electronic-Waste commented on pkg/controller.v1beta1/trial/trial_controller_util.go:
```diff
@@ -70,6 +70,13 @@ func (r *ReconcileTrial) UpdateTrialStatusCondition(instance *trialsv1beta1.Tria
+			// If the type of metrics collector is Push, We should insert an unavailable value to Katib DB
+			if instance.Spec.MetricsCollector.Collector.Kind == commonv1beta1.PushCollector {
+				if err := r.reportUnavailableMetrics(instance); err != nil {
+					logger.Error(err, "Failed to insert unavailable value to Katib DB")
```
Yeah, I agree with you since retry reconciliation is better for inserting metrics into Katib DB. I'll modify the code.

Electronic-Waste replied:
@tenzen-y The retrying reconciliation logic has been added now. PTAL👀

Electronic-Waste replied:
@andreyvelich I've added `errReportMetricsFailed` to describe the error caused by failing to report unavailable metrics to Katib DB as you suggested in the meeting today. Please let me know if there are any changes I need to make further.

johnugeorge commented:
/lgtm

tenzen-y commented:
/area gsoc

Electronic-Waste commented:
/rerun-all

Electronic-Waste commented:
/rerun-all

Electronic-Waste commented:
/rerun-all

andreyvelich commented on pkg/controller.v1beta1/trial/trial_controller.go:
```diff
@@ -57,6 +58,8 @@ var (
 	log = logf.Log.WithName(ControllerName)
 	// errMetricsNotReported is the error when Trial job is succeeded but metrics are not reported yet
 	errMetricsNotReported = fmt.Errorf("metrics are not reported yet")
+	// errReportMetricsFailed is the error when metrics are reported but not successful yet
+	errReportMetricsFailed = fmt.Errorf("failed to report metrics")
```
```suggestion
	// errReportMetricsFailed is the error when `unavailable` metrics value can't be inserted to the Katib DB.
	errReportMetricsFailed = fmt.Errorf("failed to report unavailable metrics")
```

andreyvelich replied:
@Electronic-Waste Just a small nit from my side and we can merge it.

Electronic-Waste replied:
@andreyvelich I've committed your suggestions :)

andreyvelich commented:
Thank you for this @Electronic-Waste!
/lgtm
/approve

tenzen-y reviewed (commented):
/hold

tenzen-y commented on pkg/controller.v1beta1/trial/trial_controller_util.go:
```diff
@@ -70,6 +70,15 @@ func (r *ReconcileTrial) UpdateTrialStatusCondition(instance *trialsv1beta1.Tria
+			if instance.Spec.MetricsCollector.Collector.Kind == commonv1beta1.PushCollector {
+				if err := r.reportUnavailableMetrics(instance); err != nil {
+					logger.Error(err, "Failed to insert unavailable value to Katib DB")
+					return errReportMetricsFailed
```
```suggestion
					return fmt.Errorf("%w: %w", errReportMetricsFailed, err)
```

We should wrap the error.

Electronic-Waste replied:
LGTM

tenzen-y commented on pkg/controller.v1beta1/trial/trial_controller.go:
```diff
@@ -179,7 +182,7 @@ func (r *ReconcileTrial) Reconcile(ctx context.Context, request reconcile.Reques
 		err := r.reconcileTrial(instance)
 		if err != nil {
-			if err == errMetricsNotReported {
+			if err == errMetricsNotReported || err == errReportMetricsFailed {
```
```suggestion
			if errors.Is(err, errMetricsNotReported) || errors.Is(err, errReportMetricsFailed) {
```

We should not directly compare errors.

Electronic-Waste replied:
Yeah, I agree with you.

Electronic-Waste commented:
@tenzen-y I've updated the code as you suggested. Please let me know of any modifications I need to make :)

tenzen-y commented on pkg/controller.v1beta1/trial/trial_controller_test.go:
```diff
@@ -290,14 +290,82 @@ func TestReconcileBatchJob(t *testing.T) {
+		}, timeout).Should(gomega.BeTrue())
+	})
+
+	t.Run(`Trail with "Complete" BatchJob and Unavailable metrics(Push MC).`, func(t *testing.T) {
```
https://github.com/kubeflow/katib/actions/runs/10598019897/job/29371352434?pr=2394#step:4:132

tenzen-y commented on pkg/controller.v1beta1/trial/trial_controller_test.go:
```diff
@@ -290,14 +290,82 @@ func TestReconcileBatchJob(t *testing.T) {
+		}, timeout).Should(gomega.BeTrue())
+	})
+
+	t.Run(`Trail with "Complete" BatchJob and Unavailable metrics(Push MC, failed once).`, func(t *testing.T) {
```
https://github.com/kubeflow/katib/actions/runs/10598019897/job/29371352434?pr=2394#step:4:138

tenzen-y reviewed (commented):
Some test cases are so flaky.
@Electronic-Waste Could you investigate them?

Electronic-Waste commented:
@tenzen-y Okay,I'll investigate it.

Electronic-Waste commented:
/rerun-all

Electronic-Waste commented:
/rerun-all

Electronic-Waste commented:
@tenzen-y @andreyvelich  I think it's difficult for me to figure out how to solve this error.

By adding logs to the test file, I found that the error occurred because the reconciliation is sometimes not triggered when `c.Delete` is called, thus giving rise to the failure to remove finalizers. As a result, the former Trial can't be deleted correctly and the next UT will report `object is being deleted: trials.kubeflow.org "test-trial" already exists` while creating a new Trial.

But I don't know why the deletion sometimes works and sometimes not. Could you please help me figure out the reason for it? Or we can open a new issue to discuss this :)

REF: https://github.com/kubeflow/katib/actions/runs/10635118286/job/29484008031?pr=2394
<img width="1036" alt="截屏2024-08-30 23 58 32" src="https://github.com/user-attachments/assets/082aea90-bd6d-4d1c-9dce-05d6df185829">

REF: https://github.com/kubeflow/katib/actions/runs/10634460384/job/29482355521
<img width="1031" alt="截屏2024-08-31 00 08 55" src="https://github.com/user-attachments/assets/cc1b7e68-a2f8-4965-8d12-04e52a84019d">

andreyvelich commented:
@Electronic-Waste Do you use different Trial name for each test case ?

Electronic-Waste commented:
> @Electronic-Waste Do you use different Trial name for each test case ?

@andreyvelich I use the same name for each test case. But I think it makes no difference to use different names because the failure of deleting Trial is due to the missing of reconciliation.

Electronic-Waste commented:
@tenzen-y @andreyvelich I updated the order of UT and it just turned out to be useful to resolve the flasky error.

I don't know why but it just works🤣.

andreyvelich reviewed (commented):
I think we can merge it.
/lgtm
/assign @tenzen-y 
Please take a look if that looks good to you.

andreyvelich commented:
/hold

tenzen-y commented:
> @tenzen-y @andreyvelich I updated the order of UT and it just turned out to be useful to resolve the flasky error.
> 
> I don't know why but it just works🤣. Maybe the error results from some tricky bugs in `envtest`.

@Electronic-Waste It seems that there are still flaky situations in the same test cases: 

- https://github.com/kubeflow/katib/actions/runs/10654478578/job/29576729814?pr=2394#step:4:129
- https://github.com/kubeflow/katib/actions/runs/10654478578/job/29576845852?pr=2394#step:4:127
- https://github.com/kubeflow/katib/actions/runs/10654478578/job/29576845852?pr=2394#step:4:127

1/4 trials succeeded, otherwise failed.

andreyvelich commented on pkg/controller.v1beta1/trial/trial_controller_test.go:
```diff
@@ -290,14 +290,82 @@ func TestReconcileBatchJob(t *testing.T) {
-		// Create the Trial
-		trial := newFakeTrialBatchJob()
+		// Create the Trial with Push MC
+		trial := newFakeTrialBatchJob(commonv1beta1.PushCollector)
```
Try to use different Trial name for every test case, maybe it will help.

Electronic-Waste replied:
Okay, I'll try that.

Electronic-Waste replied:
It does not work...

andreyvelich commented on pkg/controller.v1beta1/trial/trial_controller_test.go:
```diff
@@ -290,14 +290,82 @@ func TestReconcileBatchJob(t *testing.T) {
+	t.Run(`Trail with "Complete" BatchJob and Unavailable metrics(Push MC, failed once).`, func(t *testing.T) {
+		g := gomega.NewGomegaWithT(t)
+		gomock.InOrder(
+			mockManagerClient.EXPECT().GetTrialObservationLog(gomock.Any()).Return(observationLogUnavailable, nil).MinTimes(1),
+			mockManagerClient.EXPECT().ReportTrialObservationLog(gomock.Any(), gomock.Any()).Return(nil, errReportMetricsFailed).MinTimes(1),
+			mockManagerClient.EXPECT().GetTrialObservationLog(gomock.Any()).Return(observationLogUnavailable, nil).MinTimes(1),
```
Why do you verify that GetTrialObservationLog will be called twice ?

Electronic-Waste replied:
Because Trial Controller will retry reconciliation when the `ReportTrialObservationLog` func fails. And then `GetTrialObservationLog` will be called again. 

I want to verify that `GetTrialObservationLog` is called after `ReportTrialObservationLog` fails.

Electronic-Waste commented:
/rerun-all

Electronic-Waste commented:
/rerun-all

Electronic-Waste commented:
@andreyvelich @tenzen-y I've fixed the flaky error by separating the UTs and adjusting the `EXPECT()` mock clause!

AFAIK, the flaky error is caused by the uncertain triggering times of the reconciliation, thus giving rise to the uncertainty of the times we call the function. The most annoying issue was that we must call `GetObservationLog` and `ReportObservationLog` in turn and we didn't know the times they were triggered when we use Push MC. 

So I reserved some `EXPECT()` function in the `gomock.InOrder` and added some new `.AnyTimes()` clause outside it. Please let me know if there are any modifications need to be made to this PR. Thanks!

Electronic-Waste commented:
@kubeflow/wg-automl-leads it seems that the coverage reports have some accidents. Could you please re-rerun these test cases and check them again?

andreyvelich commented:
@Electronic-Waste I think, you can re-trigger tests by add this comment: `/rerun-all`

andreyvelich commented:
/rerun-all

Electronic-Waste commented:
Thanks @andreyvelich
I think what I mean is running those few failed test cases. Sorry for not realizing that I can re-run them all without you. I'll re-rerun them on my own in the future.

andreyvelich commented:
Not sure why coveralls fails tho.
@Electronic-Waste Please can you try to investigate it ?
cc @kubeflow/wg-training-leads

Electronic-Waste commented:
@andreyvelich AFAIK It fails sometimes. It may turn normal in a few hours.

Electronic-Waste commented:
/rerun-all

Electronic-Waste commented:
/rerun-all

Electronic-Waste commented:
> @andreyvelich There is an issue describing it: https://github.com/lemurheavy/coveralls-public/issues/1716
>
> It seems that the coverage report will fail if we rerun the old CI build: https://github.com/lemurheavy/coveralls-public/issues/1716#issuecomment-1593414065

@tenzen-y  FYR, the coverage report will fail if we rerun the old CI build. 

The way I can think of to examine these is to rerun all CI tests.

tenzen-y commented:
/rerun-all

Electronic-Waste commented:
@tenzen-y  The `/rerun-all` command only restarts `Go Test`. We need to restart all CI builds since the bug report is:

```
{"message":"Can't add a job to a build that is already closed. Build 10789114390 is closed. See docs.coveralls.io/parallel-builds","error":true}
```

Electronic-Waste commented:
I made some tiny changes in the comment lines. It will retrigger all CI builds:)

Maybe this can help you check the robustness of the UTs in `Go Test` @tenzen-y :)

Electronic-Waste commented:
/rerun-all

Electronic-Waste commented:
/rerun-all

Electronic-Waste commented:
> @andreyvelich @tenzen-y I've fixed the flaky error by separating the UTs and adjusting the `EXPECT()` mock clause!
> 
> 
> 
> AFAIK, the flaky error is caused by the uncertain triggering times of the reconciliation, thus giving rise to the uncertainty of the times we call the function. The most annoying issue was that we must call `GetObservationLog` and `ReportObservationLog` in turn and we didn't know the times they were triggered when we use Push MC. 
> 
> 
> 
> So I reserved some `EXPECT()` function in the `gomock.InOrder` and added some new `.AnyTimes()` clause outside it. Please let me know if there are any modifications need to be made to this PR. Thanks!

@tenzen-y I'm sure that the flaky error has been addressed now. May I ask whether you need to check it again or not? I can trigger the CI builds again by making more tiny changes since the coverage report only works when we retrigger all CI builds.

tenzen-y commented:
> > @andreyvelich @tenzen-y I've fixed the flaky error by separating the UTs and adjusting the `EXPECT()` mock clause!
> > AFAIK, the flaky error is caused by the uncertain triggering times of the reconciliation, thus giving rise to the uncertainty of the times we call the function. The most annoying issue was that we must call `GetObservationLog` and `ReportObservationLog` in turn and we didn't know the times they were triggered when we use Push MC.
> > So I reserved some `EXPECT()` function in the `gomock.InOrder` and added some new `.AnyTimes()` clause outside it. Please let me know if there are any modifications need to be made to this PR. Thanks!
> 
> @tenzen-y I'm sure that the flaky error has been addressed now. May I ask whether you need to check it again or not? I can trigger the CI builds again by pushing more tiny changes since the coverage report only works when we retrigger all CI builds.

Thank you for driving this! Throughout this 2 week CI result, I am sure that we succeeded to get rid of flakiness root causes.

https://github.com/kubeflow/katib/actions/workflows/test-go.yaml

tenzen-y commented on pkg/controller.v1beta1/trial/trial_controller_test.go:
```diff
@@ -134,59 +168,21 @@ func TestReconcileBatchJob(t *testing.T) {
-	wg := &sync.WaitGroup{}
-	wg.Add(1)
+	// Start test manager
+	mgrCtx, cancel := context.WithCancel(context.TODO())
```
```suggestion
	mgrCtx, cancel := context.WithCancel(context.TODO())
	t.Cleanup(cancel)
```

Instead of calling cancel function in line 369, could you enforce to call cancel after this test is finished?

Electronic-Waste replied:
Sure! I think it looks better to me.

tenzen-y commented on pkg/controller.v1beta1/trial/trial_controller_test.go:
```diff
@@ -326,6 +364,9 @@ func TestReconcileBatchJob(t *testing.T) {
 		g := gomega.NewGomegaWithT(t)
 		g.Expect(r.updateStatus(&trialsv1beta1.Trial{})).To(gomega.HaveOccurred())
 	})
+
+	// Stop the test manager
+	cancel()
```
```suggestion
```

tenzen-y commented on pkg/controller.v1beta1/trial/trial_controller_test.go:
```diff
@@ -290,14 +287,55 @@ func TestReconcileBatchJob(t *testing.T) {
+
+		// Expect that Trial status is succeeded with "false" status and "metrics unavailable" reason.
+		// Metrics unavailable because GetTrialObservationLog returns "unavailable".
+		g.Eventually(func() bool {
+			if err = c.Get(ctx, trialKey, trial); err != nil {
+				return false
+			}
+			return trial.IsMetricsUnavailable() &&
+				len(trial.Status.Observation.Metrics) > 0 &&
+				trial.Status.Observation.Metrics[0].Min == consts.UnavailableMetricValue &&
+				trial.Status.Observation.Metrics[0].Max == consts.UnavailableMetricValue &&
+				trial.Status.Observation.Metrics[0].Latest == consts.UnavailableMetricValue
+		}, timeout).Should(gomega.BeTrue())
```
```suggestion
		g.Eventually(func(g gomega.Gomega) {
			g.Expect(c.Get(ctx, trialKey, trial)).Should(gomega.Succeeded())
			g.Expect(trial.IsMetricsUnavailable()).Should(gomega.BeTrue)
			g.Expect(trial.Status.Observation.Metrics).ShouldNot(gomega.HaveLen(0))
			g.Expect(trial.Status.Observation.Metrics[0]).Should(gomega.BeComparableTo(commonv1beta1.Metric{
			  Name: METRCS_NAME,
			  Min: consts.UnavailableMetricValue,
			  Max: consts.UnavailableMetricValue,
			  Latest: consts.UnavailableMetricValue,
			}))
		}, timeout).Should(gomega.Succeeded())
```

Could you use the propagated gomega instance to improve debuggability?

tenzen-y replied:
I guess you just copy and paste here, but we should select the ideal approach in the new codes.
If you can replace the not ideally codes with ideal forms in followup PR, it would be great.

Electronic-Waste replied:
Yeah, I think the propagated gomega instance looks better. 
I'll raise a new issue describing it and open a PR to it later :)

tenzen-y reviewed (commented):
The most things lgtm

tenzen-y reviewed (commented):
That was great improvments!
Thank you for doing this!

/lgtm
/approve

Electronic-Waste commented:
Thank you for your detailed review @tenzen-y!

This PR is holding now. Can you remove the `hold` label so that this PR can be merged?

tenzen-y commented:
> Thank you for your detailed review @tenzen-y!
> 
> This PR is holding now. Can you remove the `hold` label so that this PR can be merged?

Sure.
/hold cancel
